File: Assets/_Scripts/States/Player/PlayerSkatingState.cs
using Styx.Entities.PlayerModule;
using RageCure.StateUtils;
using UnityEngine;
using Styx.Commons.Utils;

namespace Styx.States
{
    public class PlayerSkatingState : State<PlayerState>
    {
        private void Rotate(float turnAmount, Player player)
        {
            transform.Rotate(0f, turnAmount * player.MovementModel.RotationSpeed * Time.deltaTime, 0f);
        }

        private void Move(float forwardAmount, Player player)
        {
            player.AddVelocity(transform.forward * player.LinearSpeed * Time.deltaTime * forwardAmount);
        } 

        public override void Enter(PlayerState player)
        {
        }

        public override void Execute(PlayerState playerState)
        {
            var player = playerState.Player;
            player.Inputs.HorizontalAxis.Execute();
            player.Inputs.VerticalAxis.Execute();
        }

        public override void FixedExecute(PlayerState playerState)
        {
            var player = playerState.Player;
            var movementVector = (player.Inputs.VerticalAxis.Value * Vector3.forward + player.Inputs.HorizontalAxis.Value * Vector3.right);
            var normalizedMovementVector = Vector3Utils.NormalizeIfNot(movementVector);
            var move = transform.InverseTransformDirection(normalizedMovementVector);

            var forwardAmount = move.z;
            var turnAmount = Mathf.Atan2(move.x, move.z);

            Rotate(turnAmount, player);
            Move(forwardAmount, player);
            player.AnimatorModel.SetFloat("Forward Amount", forwardAmount);
        }

        public override void Exit(PlayerState playerState)
        {
        }

        //public void OnCollisionEnter(Collision collision)
        //{
        //    Debug.Log(collision.relativeVelocity.magnitude);
        //    if (!_isColliderActive && _playerState == null)
        //        return;

        //    if (collision.gameObject.layer == LayerMask.NameToLayer("Player") && collision.relativeVelocity.magnitude > 10f)
        //    {
        //        var playerId = collision.gameObject.GetComponent<PlayerId>().Value;
        //        var playerState = PlayerManager.Instance.GetPlayerState(playerId);
        //        if (playerState.StateMachine.CurrentState.GetType() != typeof(PlayerDashState))
        //            playerState.Reduce(Message.UpdateToKnockDown);
        //    }
        //}
    }
}

File: Assembly-FSharp/Assembly-FSharp/States/Player/PlayerSkatingState.fs
namespace Styx.States.Player

open UnityEngine

open Styx.Entities.PlayerModule
open RageCure.StateUtils
open RageCure.Commons.Utils

type PlayerSkatingState () =
    inherit State<Player> ()

    member this.Enter (player: Player) =
        ()

    member this.Execute (player: Player) =
        player.inputs.horizontalAxis.Execute()
        player.inputs.verticalAxis.Execute()

    member this.FixedExecute (player: Player) =
        let move =
            (player.inputs.verticalAxis.Value * Vector3.forward + player.inputs.horizontalAxis.Value * Vector3.right)
            |> Vector3Utils.normalizeIfNot
            |> this.transform.InverseTransformDirection
            |> Vector3Utils.projectOnPlane Vector3.up
        let forwardAmount =
            move.z
        let turnAmount =
            Mathf.Atan2 (move.x, move.z)
        this.transform.Rotate(0.f, turnAmount * player.movementModel.RotationSpeed * Time.deltaTime, 0.f)
        player.rigidbodyModel.Rigidbody.velocity <- player.rigidbodyModel.Rigidbody.velocity + ((this.transform.forward * player.movementModel.LinearSpeed * Time.deltaTime) * forwardAmount)
        player.animatorModel.SetFloat "Forward Amount" forwardAmount

    member this.Exit (player: Player) =
        ()

